src/OrbitGl/App.cpp

```diff
@@ -1068,43 +1071,91 @@ void OrbitApp::LoadModules(
-      continue;
+  const auto handle_error = [this](const ErrorMessageOr<void>& result) {
+    if (result.has_error()) {
+      error_message_callback_("Error loading symbols", result.error().message());
+      return;
     }
+  };
 
-    // TODO(177304549): [new UI] maybe come up with a better indicator whether orbit is connected
-    // than process_manager != nullptr
-    if (!absl::GetFlag(FLAGS_local) && GetProcessManager() != nullptr) {
-      LoadModuleOnRemote(module, std::move(function_hashes_to_hook),
-                         std::move(frame_track_function_hashes), symbols_path.error().message());
-      continue;
-    }
+  for (const auto& module : modules) {
+    futures.emplace_back(LoadModule(module).Then(main_thread_executor_, std::move(handle_error)));
+  }
 
-    // If no symbols are found and remote loading is not attempted.
-    SendErrorToUi("Error loading symbols",
-                  absl::StrFormat("Did not find symbols for module \"%s\": %s", module->file_path(),
-                                  symbols_path.error().message()));
-    modules_currently_loading_.erase(module->file_path());
+  return orbit_base::JoinFutures(futures);
+}
+
+orbit_base::Future<ErrorMessageOr<void>> OrbitApp::LoadModule(const ModuleData* module) {
+  return LoadModule(module->file_path(), module->build_id());
+}
+
+orbit_base::Future<ErrorMessageOr<void>> OrbitApp::LoadModule(const std::string& module_path,
+                                                              const std::string& build_id) {
+  const auto it = modules_currently_loading_.find(module_path);
+  if (it != modules_currently_loading_.end()) {
+    return it->second;
   }
+
+  const auto local_symbols_path = FindModuleLocally(module_path, build_id);
+
+  if (local_symbols_path) {
+    return LoadSymbols(local_symbols_path.value(), module_path);
+    // return orbit_base::Future<ErrorMessageOr<std::filesystem::path>>{local_symbols_path};
+  }
+
+  // TODO(177304549): [new UI] maybe come up with a better indicator whether orbit is
+  // connected than process_manager != nullptr
+  if (absl::GetFlag(FLAGS_local) || GetProcessManager() == nullptr) {
+    return orbit_base::Future<ErrorMessageOr<void>>{local_symbols_path.error()};
+  }
+
+  auto load_symbols = [this, module_path, local_error_msg = local_symbols_path.error().message()](
+                          const ErrorMessageOr<std::filesystem::path>& result)
+      -> orbit_base::Future<ErrorMessageOr<void>> {
+    if (result.has_error()) {
+      return {ErrorMessage{
+          absl::StrFormat("Did not find symbols locally or on remote for module \"%s\": %s\n%s",
+                          module_path, local_error_msg, result.error().message())}};
+    }
+
+    return LoadSymbols(result.value(), module_path);
+  };
+
+  auto load_symbols_result =
+      LoadModuleOnRemote(module_path).Then(main_thread_executor_, std::move(load_symbols));
+
+  auto final_result =
+      UnwrapFuture(load_symbols_result)
```

Comment:
I'm a bit lost here. What's `UnwrapFuture` doing? `load_symbols_result` is a `Future` containing a `Future`?

Reply:
Yes, exactly. It's a Future in a Future. That's not ideal, I know.

I already have some code in flight, that specializes `Future<ErrorMessageOr<T>>` and avoids this situation all the way. 

But for now we are still mixing the happy path with the unhappy path. As soon as that's done I can improve this part again.